Avoid memory leaks in Stream methods

szeiger opened the pull request:
`Stream.collect/collectFirst/find` kept references to the original
receiver of the call which prevented any part of the Stream from
getting garbage collected as soon as possible.

The previous implementation of `Stream.collect` claimed to avoid this
but this only worked insofar as the resulting Stream did not hold on
unnecessarily to any references to the original Stream. It still kept
a reference to the original receiver while iterating over non-matching
elements which can only be avoided with a tail-recursive
implementation.

Fixes https://github.com/scala/bug/issues/11443

szeiger committed:
Avoid memory leaks in Stream methods

`Stream.collect/collectFirst/find` kept references to the original
receiver of the call which prevented any part of the Stream from
getting garbage collected as soon as possible.

The previous implementation of `Stream.collect` claimed to avoid this
but this only worked insofar as the resulting Stream did not hold on
unnecessarily to any references to the original Stream. It still kept
a reference to the original receiver while iterating over non-matching
elements which can only be avoided with a tail-recursive
implementation.

Fixes https://github.com/scala/bug/issues/11443

szeiger commented:
Huh, that's interesting. It fails when the optimizer is enabled.

lrytz commented:
We had that before, maybe it's a different situation than the one fixed in https://github.com/scala/scala/commit/505b01c06e12c9ca230a17eba1dd9a58d17e7e0b. I can take a look.

lrytz committed:
Null out the local variable holding the result of an inlined method

The inliner replaces `return` instructions by `store`s to save the
result of an inlined method. After loading that result, the local
variable should be nulled out to preven memory leaks.

This case was overlooked in 505b01c06e12c9ca230a17eba1dd9a58d17e7e0b.

lrytz commented:
```
[error] - testJarSize
[error]   - root/testJarSize failed: The library jar is too big: 5692760 bytes.
```

nice, the test works :-)

lrytz commented:
The `@tailrec` method in `Stream` doesn't compile with dotty due to https://github.com/lampepfl/dotty/issues/7116.

szeiger commented:
I'll add `final`. I didn't even know it was supposed to work for sealed classes...

lrytz commented:
MiMa doesn't know about children of sealed classes either :-) added an exception.

szeiger commented:
Phew, `final`ly!
